One HoP ticket per person, old tickets delete

cacogen opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Right now it's too easy for people to print off a bunch of tickets, litter the hall with them and leave the HoP unable to keep up with the amount of tickets, rendering the machine useless.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

No ticket spam, means the ticket machine will see more use
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: cacogen
tweak: Can only get one HoP ticket per person unless it's emagged
tweak: Previous ticket deletes ("disperses", it's "nanite paper") when HoP calls the next customer
tweak: Reduces maximum number of tickets to 100 because the refill mechanic would never be used otherwise
del: Can no longer recycle tickets using the machine, because it allows you to artificially inflate the amount of tickets the HoP has to get through
add: Tickets vibrate (nanomachines, son) when it's your turn, so you can step away from the line if it's taking too long
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

cacogen committed:
One HoP ticket per person, old tickets deleted

SkeletalElite commented:
but now the HoP cant be evil and pull a bunch of the tickets before opening the line and then make people use tickets and slowly go through all the numbers he pre pulled to make them wait.

ExcessiveUseOfCobblestone commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -181,6 +199,8 @@
 	max_integrity = 50
 	var/saved_maptext = null
+	var/mob/living/carbon/owner = null
+	var/obj/machinery/ticket_machine/source = null
```
remove the `= null`s here

ExcessiveUseOfCobblestone commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -42,8 +50,11 @@
 		return
 	playsound(src, 'sound/misc/announce_dig.ogg', 50, 0)
-	say("Next customer, please!")
+	if(current_number > 0)
```
```DM
	if(current_number && !(obj__flags & EMAGGED))
```

ExcessiveUseOfCobblestone commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -153,6 +163,10 @@
 	if(ticket_number >= max_number)
 		to_chat(user,"Ticket supply depleted, please refill this unit with a hand labeller refill cartridge!")
 		return
+	if(user in ticket_holders)
```
```DM
if((user in ticket_holders) && !(obj_flags & EMAGGED))
```

ExcessiveUseOfCobblestone commented:
wish these tickets were fluffed as like holographic or something to make the whole "it randomly deletes from your inventory" a tad more plausible.

kriskog commented:
> 
> 
> but now the HoP cant be evil and pull a bunch of the tickets before opening the line and then make people use tickets and slowly go through all the numbers he pre pulled to make them wait.

There is nothing quite like calling out and waiting for 30 numbers that quite obviously is on the floor by the ticket machine as your queue of impatient spessmen steadily grows past your shutters. I will miss this.

cacogen commented:
>wish these tickets were fluffed as like holographic or something to make the whole "it randomly deletes from your inventory" a tad more plausible.

I know. I wish I had a sprite for this. I can't sprite at all. I'll add something to the description.

cacogen committed:
does a lot look at it

cacogen commented:
So I just fluffed them to be paper made from nanites. The HoP's button sends a signal to the nanites to disperse when the next ticket is called.

MCterra10 commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -13,11 +13,13 @@
 	var/ready = TRUE
 	var/id = "ticket_machine_default" //For buttons
+	var/list/ticket_holders = list()
+	var/list/tickets = list()
```
Travis is failing because you're calling a proc on an item in a list without a type specifically defined, this doesn't technically matter, but Travis isn't happy with it.
I can see two easy ways to fix it, either define a type for the list (`var/list/obj/tickets = list()` or whatever is required, or set another variable to hold the current ticket;
```
var/obj/current_ticket = tickets[current_number]
current_ticket.visible_message()
```

cacogen replied:
Thanks!

cacogen committed:
Pleases Travis. Doesn't try to sell the nanite paper thing too hard, who cares if the stupes don't get it

Dawson1917 commented:
We have an entire role for dealing with litter... You really want to magic away away paper scraps instead of calling the janitor?

cacogen commented:
I want to limit it to one ticket per person at a time, which means the
ticket has to be deleted before they can get another. This is the simplest,
easiest way to do it. Why do you care about something so minor? The janitor
already has plenty of mess to contend with, that they will never be able to
stay on top of the way their equipment is balanced currently.

On Sat, 10 Aug 2019 at 19:41, Dawson1917 <notifications@github.com> wrote:

> We have an entire role for dealing with litter... You really want to magic
> away away paper scraps instead of calling the janitor?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/45813?email_source=notifications&email_token=AF7NO6QW7TKET4TF4CM6CRDQDZWK3A5CNFSM4IKRUSRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4AIGRY#issuecomment-520127303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF7NO6XNJLWWL5KQNQCTPGTQDZWK3ANCNFSM4IKRUSRA>
> .
>

bgobandit commented:
idk, printing a bunch of tickets and scattering them all over the station sounds exactly like classic space station 13 shenanigans

zxaber commented:
Yeah, but it also pretty much destroys any use of the ticket machine except to create a bunch of new useless objects. If the hop has to cycle through 30 numbers between each legitimate customer, why even bother?

cacogen commented:
There's also a cooldown on the HoP's button so it takes quite awhile to
catch up.

On Sun, 11 Aug 2019 at 19:58, zxaber <notifications@github.com> wrote:

> Yeah, but it also pretty much destroys any use of the ticket machine
> except to create a bunch of new useless objects. If the hop has to cycle
> through 30 numbers between each legitimate customer, why even bother?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgstation/tgstation/pull/45813?email_source=notifications&email_token=AF7NO6VT4PWA4Y7Z3QSLNJDQD7BDPA5CNFSM4IKRUSRKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4A4BCI#issuecomment-520208521>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF7NO6XJWJXJH7BMI4Q6R3DQD7BDPANCNFSM4IKRUSRA>
> .
>

actioninja commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -13,11 +13,13 @@
-	var/max_number = 999 //To stop the text going fucky. At this point, you need to refill it.
-	var/cooldown = 50 //Small cooldown, stops the clown from immediately breaking it.
+	var/current_number = 0 //What ticket number are we currently serving?
+	var/max_number = 100 //At this point, you need to refill it.
```
update the comment

actioninja replied:
wait nvm I thought you were doing the holo ticket thing

actioninja commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -27,12 +29,19 @@
 		return
 	to_chat(user, "<span class='warning'>You overload [src]'s bureaucratic logic circuitry to its MAXIMUM setting.</span>")
-	ticket_number = rand(0,999)
+	ticket_number = rand(0,100)
```
use max_number here

cacogen replied:
done

cacogen committed:
Update ticketmachine.dm

nemvar commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -202,3 +225,8 @@
+/obj/item/ticket_machine_ticket/Destroy()
+	if(owner && source)
+		source.ticket_holders -= owner
+	..()
```
``return ..()``
also set owner and source to null and remove src from source.tickets

cacogen replied:
Cobby told me to take the nulls in an earlier review but I'll add them back. Where is the src you're talking about?

cacogen replied:
@nemvar works good now

cacogen committed:
it works fine now

cacogen commented:
@actioninja could you review this again please?

ExcessiveUseOfCobblestone commented:
:omegalul:

cacogen commented:
@excessiveuseofcobblestome
pls review

ExcessiveUseOfCobblestone commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -117,52 +131,60 @@
 	. = ..()
-	if(!ready)
-		to_chat(user,"Temporarily unable to dispense ticket, please be patient!")
+	if(!ready && (obj_flags & EMAGGED))
```
is this suppose to be OR

ExcessiveUseOfCobblestone replied:
or `!(obj_flags & EMAGGED)`?

cacogen replied:
Only need the cooldown when it's emagged because when it's emagged you can print unlimited tickets (normally it's only one per person), and in testing that leads to huge piles of burning tickets which I thought would be laggy and undesirable.

ExcessiveUseOfCobblestone replied:
You already have the cooldown incorporated for all tickets, may as well remove the emagged check.

cacogen replied:
Say there's three people each getting a ticket. Without the cooldown the second person doesn't have to wait after the first and so on. And it can't be abused because until the ticket is destroyed they can't get another one. When it's emagged you want the cooldown to prevent people from printing flaming tickets as fast as they can click (they drop to the ground automatically).

ExcessiveUseOfCobblestone replied:
ok then toss the cooldown stuff under emagging and remove the emag check above?

cacogen replied:
@ExcessiveUseOfCobblestone done

cacogen committed:
adds cooldown to emag_act

ExcessiveUseOfCobblestone commented on code/modules/paperwork/ticketmachine.dm:
```diff
@@ -161,7 +162,7 @@
 /obj/machinery/ticket_machine/attack_hand(mob/living/carbon/user)
 	. = ..()
-	if(!ready && (obj_flags & EMAGGED))
+	if(!ready)
```
toss 174 and 176 under 189

cacogen replied:
@ExcessiveUseOfCobblestone done
